### DISCUSSION TITLE
Bundling all dependencies

### DISCUSSION THREAD

**Discussion by AKuederle** [pyodide/pyodide#2463] [Q&A]
I started to play around with Pyodide, and I couldn't find any information about, if it is possible to create a single bundle with all Python dependencies that the code requires, and then serve just that to the webbrowser, so that there is no need to load any additional packages via micropip.

Thanks in advance,
Arne

**Comment by hoodmane** [comment]
Thanks for the question. Not as of yet but @rth is working on this.

**Reply by danmarshall** [reply]
hello, any updates on this?

**Reply by hoodmane** [reply]
Well there's pyodide-pack: https://github.com/pyodide/pyodide-pack
Also there's a hodge-podge of tooling around making package lock files, but it has been made by a mix of rth, ryanking, joemarshall, and me all with slightly different visions and it's a bit of a three-legged race. Maybe @ryanking13 has opinions about the best approach to this right now.

I agree that not using micropip in your actual websites is usually the way to go (unless you're making a repl or similar), so we really ought to have a coherent and well-documented story for how to do that.

**Reply by ryanking13** [reply]
Yes, I think `pyodide-pack` will be our recommended tool for package bundling in the future, but I'm not sure how useful it is right now.

For now, my recommendation is to put all your packages in `python_stdlib.zip`, ideally if all the packages you need are pure-python packages (See: https://github.com/pyodide/pyodide/discussions/4575).

Otherwise (if you use c-extension packages), we don't have a best way to prebundle packages right now. You still need to download packages at runtime. But at least you can put the names of packages when loading Pyodide to load it in advance to make loading faster:

```js
pyodide.loadPackage({ packages: [<pkg1>, <pkg2>]})
```

**Reply by pfebrer** [reply]
> I agree that not using micropip in your actual websites is usually the way to go (unless you're making a repl or similar), so we really ought to have a coherent and well-documented story for how to do that.

@hoodmane Could you elaborate on why not using `micropip` is the way to go? :) I'm trying to make my pyodide environment (loading + importing) initialize faster because it takes 20s currently, making it not very practical (although it is acceptable for some use cases).

I was also wondering if one could distribute an application with an initialized environment, because I always load an import the same modules. But seeing https://github.com/pyodide/pyodide/issues/2662 I get the impression that it is not possible (?).

**Reply by pfebrer** [reply]
If you do:

```javascript
const promise = pyodide.loadPyodide({packages: ["micropip", ... long time loading packages]})
```

The promise isn't resolved until all packages are loaded, right? So you can't save time if you need to use `micropip` (which otherwise would be used to install packages asynchronously at the same time `loadPackage` is loading other dependencies).

**Reply by hoodmane** [reply]
I have been exploring memory snapshots which indeed could take some of the time off, particularly the import-time package initialization time. We recommend if you always have the same set of dependencies that you make a custom lock file that includes all of your dependencies so that all of them can be downloaded at the same time.

**Reply by pfebrer** [reply]
That's cool that you are playing with the memory snapshots :) In my case the loading phase and the importing phase take a similar amount of time, so it would be very useful!

What do you mean exactly by a "lock" file? Is there a standard way to do that?

**Reply by hoodmane** [reply]
I think the easiest way to get the lockfile is with `micropip.freeze()`. If you use Pyodide in node with NodeFS or in Chrome with NativeFS then you can persist the generated lockfile as a file. Serve it with your other files and pass the URL as the `lockFileURL` argument to `loadPyodide` and you can use `pyodide.loadPackage()` to load all of your packages in one go.

**Reply by hoodmane** [reply]
On the memory snapshot front I have figured out pretty well how the implementation has to go but there are many little API details to work out. The first part of the work is here:
https://github.com/pyodide/pyodide/pull/4652
I've been doing the snapshots for my job putting Pyodide into the Cloudflare worker runtime. There's a discussion here (mixed in with some advertising and other stuff):
https://blog.cloudflare.com/python-workers
